hazelcast/src/main/java/com/hazelcast/monitor/impl/PartitionIndexStats.java

```diff
@@ -0,0 +1,292 @@
+
+    // Per-operation stats may be safely reused/shared for operations on
+    // partitioned indexes since we know for sure only a single thread may
+    // mutate the index.
+    private final PartitionIndexOperationStats operationStats = new PartitionIndexOperationStats();
+
+    private final long creationTime;
+
+    private volatile long entryCount;
+    private volatile long queryCount;
+    private volatile long hitCount;
+    private volatile long totalHitLatency;
+    private volatile long totalNormalizedHitCardinality = Double.doubleToRawLongBits(0.0);
+    private volatile long insertCount;
+    private volatile long totalInsertLatency;
+    private volatile long updateCount;
+    private volatile long totalUpdateLatency;
+    private volatile long removeCount;
+    private volatile long totalRemoveLatency;
+    private volatile long memoryCost;
+
+    private boolean hasQueries;
+
+    public PartitionIndexStats() {
+        this.creationTime = Clock.currentTimeMillis();
+    }
+
+    private void updateMemoryCost(long delta) {
+        MEMORY_COST.lazySet(this, memoryCost + delta);
+    }
+
+    private void resetMemoryCost() {
+        MEMORY_COST.lazySet(PartitionIndexStats.this, 0);
+    }
+
+    @Override
+    public long makeTimestamp() {
+        return System.nanoTime();
+    }
+
+    @Override
+    public long getCreationTime() {
+        return creationTime;
+    }
+
+    @Override
+    public long getQueryCount() {
+        return queryCount;
+    }
+
+    @Override
+    public void incrementQueryCount() {
+        if (hasQueries) {
+            QUERY_COUNT.lazySet(this, queryCount + 1);
+        }
+    }
+
+    @Override
+    public long getHitCount() {
+        return hitCount;
+    }
+
+    @Override
+    public long getTotalHitLatency() {
+        return totalHitLatency;
+    }
+
+    @Override
+    public double getTotalNormalizedHitCardinality() {
+        return Double.longBitsToDouble(totalNormalizedHitCardinality);
+    }
+
+    @Override
+    public long getInsertCount() {
+        return insertCount;
+    }
+
+    @Override
+    public long getTotalInsertLatency() {
+        return totalInsertLatency;
+    }
+
+    @Override
+    public long getUpdateCount() {
+        return updateCount;
+    }
+
+    @Override
+    public long getTotalUpdateLatency() {
+        return totalUpdateLatency;
+    }
+
+    @Override
+    public long getRemoveCount() {
+        return removeCount;
+    }
+
+    @Override
+    public long getTotalRemoveLatency() {
+        return totalRemoveLatency;
+    }
+
+    @Override
+    public long getMemoryCost() {
+        return memoryCost;
+    }
+
+    @Override
+    public void onInsert(long timestamp, IndexOperationStats operationStats, Index.OperationSource operationSource) {
+        if (operationStats.getEntryCountDelta() == 0) {
+            // no entries were inserted
+            return;
+        }
+
+        if (operationSource == Index.OperationSource.USER) {
```

Comment:
why is the operation-source only user? Isn't it good to have these stats for all types of source?

Reply:
The idea is to avoid counting insertions/removals multiple times if they there caused by system events like partition migration. Consider a partition jumping from a member back and forth, after some time index stats will be so skewed that it will not provide any useful information to a user. To me it feels like users are expecting to pay an insertion/removal cost only at the time they are actually performing the insertion/removal.

Reply:
i see, your point makes sense. Maybe later, we can separate the stats for user originated and system originated.